src/utils/include/utils/tensor.hpp

```diff
@@ -0,0 +1,128 @@
+#ifndef TENSOR_HPP
+#define TENSOR_HPP
+
+#include <algorithm>
+#include <array>
+#include <numeric>
+#include <vector>
+
+#include <boost/range/algorithm/transform.hpp>
+#include <boost/range/numeric.hpp>
+
+namespace Utils {
+
+/**
+ * @brief Heap-allocated tensor class.
+ * @tparam T data type
+ */
+template <typename T> class Tensor {
+public:
+  using value_type = T;
+  using reference = T &;
+  using const_reference = T const &;
+  using pointer = T *;
+  using const_pointer = const value_type *;
+  using iterator = typename std::vector<T>::iterator;
+  using const_iterator = typename std::vector<T>::const_iterator;
+
+  /**
+   * @brief Creates a tensor with given extents
+   * @param extents Size for each dimension
+   */
+  explicit Tensor(std::initializer_list<std::size_t> const &extents)
+      : Tensor(std::begin(extents), std::end(extents)) {}
+
+  template <typename Iter>
+  Tensor(Iter extents_begin, Iter extents_end)
+      : m_extents(extents_begin, extents_end) {
+    m_data = std::vector<T>(std::accumulate(m_extents.begin(), m_extents.end(),
+                                            1, std::multiplies<std::size_t>()));
+
+    boost::transform(m_extents, std::back_inserter(m_strides),
+                     [stride = std::size_t(1)](std::size_t dim) mutable {
+                       auto const old = stride;
+                       stride *= dim;
+
+                       return old;
+                     });
+  }
+
+  /**
+   * @brief Element access.
+   * @param indices index for each dimension
+   */
+  template <typename... size_ts>
+  const_reference operator()(size_ts... indices) const {
+    return operator()({static_cast<std::size_t>(indices)...});
+  }
+
+  template <typename... size_ts> reference operator()(size_ts... indices) {
+    return operator()({static_cast<std::size_t>(indices)...});
+  }
+
+  const_reference
+  operator()(std::initializer_list<std::size_t> const &indices) const {
```

Comment:
I think the interface can be made clearer here if `operator[]` is used for single argument access instead. This is also hat the proposed `mdspan` and `boost::multi_array` does. E.g. use a `const_reference operator[](initializer_list<...>)` or similar. Also I'm not sure that explicitly using initializer_list here as the only type with an explicit overload leads to awkward code if you already have an array: you either have to construct an `initializer_list`, which is just noise, or explicitly destructure it, which is complicated if you don't know the rank statically. You could either template the type, or at least use an explicit version with `Utils::Span<const size_t>` which will implicitly construct from all containers that have `.data()` and `.size()`. In general I think it is also not common client code that the indices are literals, so that the curly braces syntax isn't really used that often (only looking at unit tests may be a bit misleading here).

Reply:
> To provide multidimensional array access semantics, e.g. to implement a 3D array access a[i][j][k] = x;, operator[] has to return a reference to a 2D plane, which has to have its own operator[] which returns a reference to a 1D row, which has to have operator[] which returns a reference to the element. To avoid this complexity, some libraries opt for overloading operator() instead, so that 3D access expressions have the Fortran-like syntax a(i, j, k) = x;

from https://en.cppreference.com/w/cpp/language/operators